File: impl/c/norvig.c
/* Norvig-style constraint propagation + search (bitmask candidates).
 *
 * Port of impl/python/norvig.py: eliminate + naked singles + hidden
 * singles to a fixpoint, then depth-first search with MRV cell choice.
 * Single-threaded; the `threads` argument is accepted and ignored.
 */
#include <inttypes.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define ALL 0x3FEu /* bits 1..9 set */

static uint8_t PEERS[81][20];   /* 20 peers per cell */
static uint8_t UNITS[27][9];    /* 9 rows, 9 cols, 9 boxes */
static uint8_t UNIT_OF[81][3];  /* the 3 units containing each cell */

static void init_tables(void)
{
    for (int r = 0; r < 9; r++)
        for (int c = 0; c < 9; c++) {
            UNITS[r][c] = (uint8_t)(r * 9 + c);
            UNITS[9 + c][r] = (uint8_t)(r * 9 + c);
        }
    for (int b = 0; b < 9; b++) {
        int br = b / 3, bc = b % 3;
        for (int k = 0; k < 9; k++)
            UNITS[18 + b][k] =
                (uint8_t)((br * 3 + k / 3) * 9 + bc * 3 + k % 3);
    }
    for (int i = 0; i < 81; i++) {
        UNIT_OF[i][0] = (uint8_t)(i / 9);
        UNIT_OF[i][1] = (uint8_t)(9 + i % 9);
        UNIT_OF[i][2] = (uint8_t)(18 + (i / 27) * 3 + (i % 9) / 3);
        bool seen[81] = { false };
        seen[i] = true;
        int np = 0;
        for (int t = 0; t < 3; t++)
            for (int k = 0; k < 9; k++) {
                int j = UNITS[UNIT_OF[i][t]][k];
                if (!seen[j]) {
                    seen[j] = true;
                    PEERS[i][np++] = (uint8_t)j;
                }
            }
    }
}

static bool assign(uint16_t *cand, int i, uint16_t bit);

static bool eliminate(uint16_t *cand, int i, uint16_t bit)
{
    uint16_t c = cand[i];
    if (!(c & bit))
        return true;
    c &= (uint16_t)~bit;
    if (!c)
        return false;
    cand[i] = c;
    if (!(c & (c - 1))) {               /* naked single: strip from peers */
        for (int k = 0; k < 20; k++) {
            int p = PEERS[i][k];
            if ((cand[p] & c) && !eliminate(cand, p, c))
                return false;
        }
    }
    for (int t = 0; t < 3; t++) {       /* hidden single for `bit`? */
        const uint8_t *u = UNITS[UNIT_OF[i][t]];
        int place = -1;
        for (int k = 0; k < 9; k++) {
            if (cand[u[k]] & bit) {
                if (place >= 0) {
                    place = -2;
                    break;
                }
                place = u[k];
            }
        }
        if (place == -1)                /* bit has nowhere left in unit */
            return false;
        if (place >= 0) {
            uint16_t pc = cand[place];
            if ((pc & (pc - 1)) && !assign(cand, place, bit))
                return false;
        }
    }
    return true;
}

static bool assign(uint16_t *cand, int i, uint16_t bit)
{
    uint16_t other = cand[i] & (uint16_t)~bit;
    while (other) {
        uint16_t lb = (uint16_t)(other & -other);
        other ^= lb;
        if (!eliminate(cand, i, lb))
            return false;
    }
    return true;
}

/* On success, cand holds the solved grid (all cells single bits). */
static bool search(uint16_t *cand)
{
    int best = -1, best_n = 10;
    for (int i = 0; i < 81; i++) {      /* MRV cell */
        uint16_t c = cand[i];
        if (c & (c - 1)) {
            int n = __builtin_popcount(c);
            if (n < best_n) {
                best = i;
                best_n = n;
                if (n == 2)
                    break;
            }
        }
    }
    if (best < 0)
        return true;                    /* all cells singles: solved */
    uint16_t c = cand[best];
    while (c) {
        uint16_t bit = (uint16_t)(c & -c);
        c ^= bit;
        uint16_t trial[81];
        memcpy(trial, cand, sizeof trial);
        if (assign(trial, best, bit) && search(trial)) {
            memcpy(cand, trial, sizeof trial);
            return true;
        }
    }
    return false;
}

static bool solve(const uint8_t *grid, uint16_t *cand)
{
    for (int i = 0; i < 81; i++)
        cand[i] = ALL;
    for (int i = 0; i < 81; i++) {
        int d = grid[i];
        if (d) {
            uint16_t bit = (uint16_t)(1u << d);
            if (cand[i] != bit && !assign(cand, i, bit))
                return false;
        }
    }
    return search(cand);
}

int main(int argc, char **argv)
{
    long reps = 1;
    if (argc > 1) {
        reps = strtol(argv[1], NULL, 10);
        if (reps < 1)
            reps = 1;
    }
    /* argv[2] (threads) intentionally ignored: single-threaded solver */

    init_tables();

    size_t n = 0, cap = 0;
    uint8_t (*grids)[81] = NULL;
    char *line = NULL;
    size_t linecap = 0;
    ssize_t len;
    while ((len = getline(&line, &linecap, stdin)) != -1) {
        while (len > 0 && (line[len - 1] == '\n' || line[len - 1] == '\r'))
            line[--len] = '\0';
        if (len == 0 || line[0] == '#')
            continue;
        if (n == cap) {
            cap = cap ? cap * 2 : 256;
            grids = realloc(grids, cap * sizeof *grids);
            if (!grids)
                return 0;
        }
        bool valid = (len == 81);
        for (int i = 0; valid && i < 81; i++)
            valid = (line[i] >= '0' && line[i] <= '9');
        if (valid) {
            for (int i = 0; i < 81; i++)
                grids[n][i] = (uint8_t)(line[i] - '0');
        } else {
            memset(grids[n], 10, 81);   /* impossible clue -> UNSOLVED */
        }
        n++;
    }
    free(line);

    uint16_t (*cands)[81] = malloc((n ? n : 1) * sizeof *cands);
    bool *ok = malloc((n ? n : 1) * sizeof *ok);
    if (!cands || !ok)
        return 0;

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (long r = 0; r < reps; r++)
        for (size_t i = 0; i < n; i++)
            ok[i] = solve(grids[i], cands[i]);
    clock_gettime(CLOCK_MONOTONIC, &t1);

    for (size_t i = 0; i < n; i++) {
        if (ok[i]) {
            char out[82];
            for (int j = 0; j < 81; j++)
                out[j] = (char)('0' + __builtin_ctz(cands[i][j]));
            out[81] = '\0';
            puts(out);
        } else {
            puts("UNSOLVED");
        }
    }

    uint64_t ns = (uint64_t)(t1.tv_sec - t0.tv_sec) * 1000000000u
                + (uint64_t)(t1.tv_nsec - t0.tv_nsec);
    fprintf(stderr, "ns=%" PRIu64 "\n", ns);
    free(grids);
    free(cands);
    free(ok);
    return 0;
}

File: impl/rust/src/bin/norvig.rs
//! Norvig-style constraint propagation + search (bitmask candidates).
//!
//! Eliminate + naked singles + hidden singles to a fixpoint, then
//! depth-first search with MRV cell choice.

use sudoku::{parse_args, read_puzzles, run_solver, write_results, Grid};

const ALL: u16 = 0x3FE; // bits 1..9 set

struct Tables {
    peers: [[u8; 20]; 81],   // 20 peers per cell
    units: [[u8; 9]; 27],    // 9 rows, 9 cols, 9 boxes
    unit_of: [[u8; 3]; 81],  // the 3 units containing each cell
}

fn init_tables() -> Tables {
    let mut t = Tables {
        peers: [[0; 20]; 81],
        units: [[0; 9]; 27],
        unit_of: [[0; 3]; 81],
    };
    for r in 0..9 {
        for c in 0..9 {
            t.units[r][c] = (r * 9 + c) as u8;
            t.units[9 + c][r] = (r * 9 + c) as u8;
        }
    }
    for b in 0..9 {
        let (br, bc) = (b / 3, b % 3);
        for k in 0..9 {
            t.units[18 + b][k] = ((br * 3 + k / 3) * 9 + bc * 3 + k % 3) as u8;
        }
    }
    for i in 0..81 {
        t.unit_of[i] = [
            (i / 9) as u8,
            (9 + i % 9) as u8,
            (18 + (i / 27) * 3 + (i % 9) / 3) as u8,
        ];
        let mut seen = [false; 81];
        seen[i] = true;
        let mut np = 0;
        for u in 0..3 {
            for k in 0..9 {
                let j = t.units[t.unit_of[i][u] as usize][k] as usize;
                if !seen[j] {
                    seen[j] = true;
                    t.peers[i][np] = j as u8;
                    np += 1;
                }
            }
        }
    }
    t
}

fn eliminate(t: &Tables, cand: &mut [u16; 81], i: usize, bit: u16) -> bool {
    let mut c = cand[i];
    if c & bit == 0 {
        return true;
    }
    c &= !bit;
    if c == 0 {
        return false;
    }
    cand[i] = c;
    if c & (c - 1) == 0 {
        // naked single: strip from peers
        for k in 0..20 {
            let p = t.peers[i][k] as usize;
            if cand[p] & c != 0 && !eliminate(t, cand, p, c) {
                return false;
            }
        }
    }
    for u in 0..3 {
        // hidden single for `bit`?
        let unit = &t.units[t.unit_of[i][u] as usize];
        let mut place: i32 = -1;
        for &cell in unit {
            if cand[cell as usize] & bit != 0 {
                if place >= 0 {
                    place = -2;
                    break;
                }
                place = cell as i32;
            }
        }
        if place == -1 {
            return false; // bit has nowhere left in unit
        }
        if place >= 0 {
            let p = place as usize;
            let pc = cand[p];
            if pc & (pc - 1) != 0 && !assign(t, cand, p, bit) {
                return false;
            }
        }
    }
    true
}

fn assign(t: &Tables, cand: &mut [u16; 81], i: usize, bit: u16) -> bool {
    let mut other = cand[i] & !bit;
    while other != 0 {
        let lb = other & other.wrapping_neg();
        other ^= lb;
        if !eliminate(t, cand, i, lb) {
            return false;
        }
    }
    true
}

/// On success, cand holds the solved grid (all cells single bits).
fn search(t: &Tables, cand: &mut [u16; 81]) -> bool {
    let mut best = usize::MAX;
    let mut best_n = 10;
    for i in 0..81 {
        // MRV cell
        let c = cand[i];
        if c & (c - 1) != 0 {
            let n = c.count_ones();
            if n < best_n {
                best = i;
                best_n = n;
                if n == 2 {
                    break;
                }
            }
        }
    }
    if best == usize::MAX {
        return true; // all cells singles: solved
    }
    let mut c = cand[best];
    while c != 0 {
        let bit = c & c.wrapping_neg();
        c ^= bit;
        let mut trial = *cand;
        if assign(t, &mut trial, best, bit) && search(t, &mut trial) {
            *cand = trial;
            return true;
        }
    }
    false
}

fn solve(t: &Tables, g: &mut Grid) -> bool {
    let mut cand = [ALL; 81];
    for i in 0..81 {
        let d = g[i];
        if d != 0 {
            let bit = 1u16 << d;
            if cand[i] != bit && !assign(t, &mut cand, i, bit) {
                return false;
            }
        }
    }
    if !search(t, &mut cand) {
        return false;
    }
    for (cell, &c) in g.iter_mut().zip(cand.iter()) {
        *cell = c.trailing_zeros() as u8;
    }
    true
}

fn main() {
    let (reps, _threads) = parse_args();
    let tables = init_tables();
    let puzzles = read_puzzles();
    let (results, ns) = run_solver(reps, &puzzles, |g| solve(&tables, g));
    write_results(&results, ns);
}
